src/utils/ucc_coll_utils.c

```diff
@@ -94,6 +94,172 @@ ucc_coll_args_is_mem_symmetric(const ucc_coll_args_t *args,
+    {
+        ucc_memory_type_t mem_type = args->dst.info.mem_type;
+        memcpy(&save_info->old_asymmetric_buffer.info_v,
+               &args->src.info_v, sizeof(ucc_coll_buffer_info_v_t));
+        status = ucc_mc_alloc(&save_info->scratch,
+                              ucc_coll_args_get_total_count(args,
+                                  args->src.info_v.counts, team->size),
+                              mem_type);
+        if (ucc_unlikely(UCC_OK != status)) {
+            ucc_error("failed to allocate replacement "
+                      "memory for asymmetric buffer");
+            return status;
+        }
+        args->src.info_v.buffer = save_info->scratch->addr;
+        args->src.info_v.mem_type = mem_type;
+        return UCC_OK;
+    }
+    default:
+        break;
+    }
+    return UCC_ERR_INVALID_PARAM;
+}
+
+ucc_status_t
+ucc_coll_args_free_asymmetric_buffer(ucc_coll_task_t *task)
+{
+    ucc_status_t status = UCC_OK;
+    ucc_buffer_info_asymmetric_memtype_t *save = &task->bargs.asymmetric_save_info;
+
+    if (UCC_IS_INPLACE(task->bargs.args)) {
+        return UCC_ERR_INVALID_PARAM;
+    }
+
+    if (save->scratch == NULL) {
+        ucc_error("failure trying to free NULL asymmetric buffer");
+    }
+
+    status = ucc_mc_free(save->scratch);
+    if (ucc_unlikely(status != UCC_OK)) {
+        ucc_error("error freeing scratch asymmetric buffer: %s",
+                    ucc_status_string(status));
+    }
+    save->scratch = NULL;
+
+    return status;
+}
+
+ucc_status_t ucc_copy_asymmetric_buffer(ucc_coll_task_t *task)
+{
+    ucc_status_t                          status    = UCC_OK;
+    ucc_coll_args_t                      *coll_args = &task->bargs.args;
+    ucc_buffer_info_asymmetric_memtype_t *save      = &task->bargs.asymmetric_save_info;
+    ucc_rank_t                            size      = task->team->params.size;
+
+    if(task->bargs.args.coll_type == UCC_COLL_TYPE_SCATTERV) {
+        // copy in
+        status = ucc_mc_memcpy(save->scratch->addr,
+                        save->old_asymmetric_buffer.info_v.buffer,
+                        ucc_coll_args_get_total_count(coll_args,
+                            coll_args->src.info_v.counts, size),
+                        save->scratch->mt,
+                        save->old_asymmetric_buffer.info_v.mem_type);
+    } else if(task->bargs.args.coll_type == UCC_COLL_TYPE_SCATTER) {
+        // copy in
+        status = ucc_mc_memcpy(save->scratch->addr,
+                        save->old_asymmetric_buffer.info.buffer,
+                        ucc_dt_size(save->old_asymmetric_buffer.info.datatype) *
+                            save->old_asymmetric_buffer.info.count,
+                        save->scratch->mt,
+                        save->old_asymmetric_buffer.info.mem_type);
+    } else if(task->bargs.args.coll_type == UCC_COLL_TYPE_GATHERV) {
+        // copy out
+        status = ucc_mc_memcpy(save->old_asymmetric_buffer.info_v.buffer,
```

Comment:
here we assume that destination buffer is contig, however I don't think it's always true. For example MPI spec doesn't say anything about displacements so I think it general case it won't work.
Example
```
#rank0
args.root = 0
args.src.info.count = 1
args.dst.info_v.counts= [1, 1]
args.dst.info_v.displacements = [1, 5]

#rank1:
args.root = 0
args.src.info.count = 1
```

Reply:
Hi Sergey, I updated this so that it can handle this case. My only remaining question is, is it ok that we copy the blank space in between `info_v.buf + displacements[0]` and `info_v.buf + displacements[1]`? The way I updated this PR will handle it by copying this memory.

I am assuming this works because the MPI standard defines these displacements as an offset from the buffer. The standard implies a negative stride is invalid, and also, every example allocates this empty space.